src/kOS/SharedObjects.cs

```diff
@@ -19,14 +19,20 @@ public class SharedObjects : Safe.SafeSharedObjects
         public SharedObjects()
         {
             AllVoiceValues = new Dictionary<int, VoiceValue>();
-            GameEvents.onVesselDestroy.Add(OnVesselDestroy);
         }
 
-        private void OnVesselDestroy(Vessel data)
+        public void DestroyObjects()
         {
-            if (data.id == Vessel.id)
+            if (BindingMgr != null) { BindingMgr.Dispose(); }
+            if (Window != null) { UnityEngine.Object.Destroy(Window); }
+            if (SoundMaker != null) { SoundMaker.StopAllVoices(); }
+            var props = typeof(SharedObjects).GetProperties();
+            foreach (var prop in props)
             {
-                BindingMgr.Dispose();
+                if (!prop.PropertyType.IsValueType)
+                {
+                    prop.SetValue(this, null, null);
+                }
```

Comment:
Honestly, I think that every time we attempt to access a property on shared it should be null checked.  But you're right, the bigger issue is that we could potentially null check before another thread sets the value to null.  Since I'm only setting the reference to null, you can easily get around this by storing the reference locally in the other thread object.  The calls to `Dispose` (which existed before my change) are actually a bigger deal because they potentially release unmanaged resources and set internal properties to null.